batch2/batch/driver/instance.py

```diff
@@ -105,6 +105,15 @@ def state(self):
         self._state = 'deleted'
         self.instance_pool.adjust_for_add_instance(self)
 
+    async def mark_jobs_ended(self, timestamp, reason):
```

Comment:
I think you can remove this and fold the behavior into `deactivate_instance`.  This also feels better: we should always be setting the end_time and reason when we're updating it because we know it isn't running anymore (or is about to stop running as the case may be).

Reply:
This is a bit difficult. There's a lot of things that call deactivate_instance. I can try and add it with a default reason of 'unknown' and timestamp = now unless the reason and timestamp have been explicitly passed to the function.

Reply:
I feel like the cases where we deactivate/unschedule but don't set the end time are actually bugs.

Yes.  But aren't those the real reasons?  I say it should take a reason and an optional timestamp (and use now if it isn't passed).  I realize you might need to pass the reason down a level or two to get the proper reason.  As you say, the metrics on these are going to be interesting.

Reply:
I'm worried that if we update the timestamp and reason for everything (not just explicitly for preemption/deletion), then it might break my logic where the job could continue running after the instance is marked deactivated (temporary network failure attempting to contact instance) and we won't get the true end time. I'm thinking the defaults should just be None for both reason and timestamp. I can look at where we call deactivate more closely though.

Reply:
Break your logic where?  In that case, we'll just take the earlier time.  I originally argued for taking the true end time based on the reason, but basically I'm convinced this is super rare and the difference isn't worth the complexity.

Reply:
I think you're right and this is fine. I added a bunch of reasons, but I don't think we should see any reasons in the attempts table after all the gevents have been processed except 'preempted' and 'deleted'.